Docsp-28793: count fundamentals (#103)
mongodb/docs-csharp: source/fundamentals/crud/read-operations/count.txt

   .. output::
      :language: none
      :visible: false

      Estimated number of documents in the students collection: 6

.. _csharp-count-aggregation:

Aggregation
-----------

You can use the ``Count()`` builder method to count the number
of documents in an aggregation pipeline.

Example
~~~~~~~

The following example performs the following actions:

- Specifies a match stage to find documents with a ``FinalGrade`` value
  greater than ``80``
- Counts the number of documents that match the criteria

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var filter = Builders<Student>
        .Filter.Gt(s => s.FinalGrade, 80);
      var result = _myColl.Aggregate().Match(filter).Count();
      Console.WriteLine("Number of documents with a final grade more than 80: " + result.First().Count);

   .. output::
      :language: none
      :visible: false

      Number of documents with a final grade more than 80: 2
      

Additional Information
----------------------

To learn more about the operations mentioned, see the following
guides:

- :ref:`csharp-specify-query`
- :ref:`csharp-bson`
- :ref:`csharp-guids`
- :ref:`csharp-builders` 

> - :ref:`csharp-poco`

Reviewer: Suggestion: add a link to the [$count documentation](https://www.mongodb.com/docs/manual/reference/operator/aggregation/count/) in the Server manual

collection.

.. literalinclude:: /includes/fundamentals/code-examples/CountDocuments.cs
   :start-after: start-student-struct
   :end-before: end-student-struct
   :language: csharp
   :dedent:

.. note::

   The documents in the ``students`` collection use the camel-case naming
   convention. The examples in this guide use a ``ConventionPack``
   to deserialize the fields in the collection into Pascal case and map them to
   the properties in the ``Student`` class. 

   To learn more about custom serialization, see
   :ref:`csharp-custom-serialization`. 
   
.. _csharp-accurate-count:

Accurate Count
--------------

To count the number of documents that match your :ref:`query filter <csharp-specify-query>`, use the
``CountDocuments()`` method. If you pass an empty query filter, this method
returns the total number of documents in the collection.

Example 
~~~~~~~

The following example counts the number of documents where the
value of ``finalGrade`` is less than ``80``:

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp
      :dedent:

      var filter = Builders<Student>.Filter.Lt(s => s.FinalGrade, 80.0);
	  var count = _myColl.CountDocuments(filter);

> 	  Console.WriteLine("Number of documents with a final grade less than 80: " + count);

Reviewer: I: fix indentation formatting

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``Collation``
     - | The type of language collation to use when sorting results.  
       | Default: ``nil``

   * - ``Hint`` 
     - | The index to use to scan for documents to count. 
       | Default: ``nil``

   * - ``Limit`` 
     - | The maximum number of documents to count. 
       | Default: ``0`` 

   * - ``MaxTime``
     - | The maximum amount of time that the query can run on the server.
       | Default: ``nil``

   * - ``Skip`` 
     - | The number of documents to skip before counting.
       | Default: ``0``

.. tip::

   When you use ``CountDocuments()`` to return the total number of documents in a
   collection, MongoDB performs a collection scan. You can avoid a collection scan and
   improve the performance of this method by using a hint to take advantage of the built-in index on
   the ``_id`` field. Use this technique only when calling ``CountDocuments()``
   with an empty query parameter.

   .. code-block:: csharp
      :emphasize-lines: 1, 2

      var filter = Builders<Student>.Filter.Empty;
      CountOptions opts = new CountOptions(){Hint = "_id_"};

>       var counter = collection.CountDocuments(filter, opts);

Reviewer: ```suggestion
      var count = collection.CountDocuments(filter, opts);
```

.. _csharp-estimated-count:

Estimated Count
---------------

To estimate the total number of documents in your collection, use the
``EstimatedDocumentCount()`` method. 

.. note:: 

>     The ``EstimatedDocumentCount()`` method is quicker than the

Reviewer: ```suggestion
    The ``EstimatedDocumentCount()`` method is more efficient than the
```

    ``CountDocuments()`` method because it uses the collection's
    metadata rather than scanning the entire collection. 

Modify Behavior
~~~~~~~~~~~~~~~

You can modify the behavior of ``EstimatedDocumentCount()`` by passing a
``EstimatedDocumentCountOptions`` type as a parameter. If you don't
specify any options, the driver uses default values.

You can set the following properties in a ``EstimatedDocumentCountOptions`` object:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``MaxTime``
     - | The maximum amount of time that the query can run on the server.
       | Default: ``nil``

Example
```````

The following example estimates the number of documents in the
``students`` collection:

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp
      :dedent:

     var count = _myColl.EstimatedDocumentCount();
     Console.WriteLine("Estimated number of documents in the students collection: " + count);

   .. output::
      :language: none
      :visible: false

>       Estimated number of documents in the students collection: 6

Reviewer: I: example is not showing up in staging because of indentation errors
S: fix errors (text should be aligned with directive)

.. _csharp-count-aggregation:

Aggregation
-----------

> You can also include the ``Count()`` builder method to count the number

Reviewer: ```suggestion
You can use the ``Count()`` builder method to count the number
```

of documents in an aggregation pipeline.

Example
~~~~~~~

The following example performs the following actions:

> - Specifies a filter in the match stage
> - Appends a count stage using ``Count()``

Reviewer: S: describe what is happening in each stage

```suggestion
- Specifies a match stage to find documents with a ``FinalGrade`` value greater than ``80``
- Counts the number of documents that match the criteria
```

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

>      var matchStage = Builders<Student>
> 		.Filter.Gt(s => s.FinalGrade, 80);
>      var result = _myColl.Aggregate().Match(matchStage).Count();
> 	 Console.WriteLine("Number of documents with a final grade more than 80: " + result.First().Count);

Reviewer: I: indentation error, also, these variables should be renamed correctly

```suggestion
     var filter = Builders<Student>
		.Filter.Gt(s => s.FinalGrade, 80);
     var result = _myColl.Aggregate().Match(filter).Count();
     Console.WriteLine("Number of documents with a final grade more than 80: " + result.First().Count);
```

The following ``Student`` class is a Plain Old CLR/Class Object, or
:ref:`poco <csharp-poco>`,  that models the documents in this
collection.

.. literalinclude:: /includes/code-examples/count-documents/Student.cs
   :start-after: start-grades-struct
   :end-before: end-grades-struct
   :language: csharp
   :dedent:

.. note::

   The documents in the ``grades`` collection use the camel-case naming
   convention. The examples in this guide use a ``ConventionPack``
   to deserialize the fields in the collection into Pascal case and map them to
   the properties in the ``Student`` class. 

   To learn more about custom serialization, see
   :ref:`csharp-custom-serialization`. 
   
.. _csharp-accurate-count:

Accurate Count
--------------

To count the number of documents that match your :ref:`query filter <csharp-specify-query>`, use the
``CountDocuments()`` method. If you pass an empty query filter, this method
returns the total number of documents in the collection.

Example 
~~~~~~~

The following example counts the number of documents where the
``finalGrade`` is less than ``80``:

.. io-code-block::
   :copyable: true

   .. input::

>       :language: csharp

Reviewer: S: let's create a code file and use includes here instead of copy pasting code. We can discuss synchronously

      var filter = Builders<BsonDocument>.Filter.Lt("finalGrade", 80);
      var count = coll.CountDocuments(filter);
      
      Console.WriteLine("Number of documents with a final grade less
      than 80: " + count);

   .. output::
      :language: none
      :visible: false

      Number of documents with a final grade less than 80: 2

Modify Behavior
~~~~~~~~~~~~~~~

> You can modify the behavior of ``CountDocuments()`` by passing in a
> ``CountOptions`` type. If you don't specify any options, the driver uses
> its default values.

Reviewer: ```suggestion
You can modify the behavior of ``CountDocuments()`` by passing a ``CountOptions`` type as a parameter. If you don't specify any options, the driver uses default values.
```

> The ``CountOptions`` type allows you to configure options with the
> following properties:

Reviewer: ```suggestion
You can set the following properties in a ``CountOptions`` object:
```

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``Collation``
     - | The type of language collation to use when sorting results.  
       | Default: ``nil``

   * - ``Hint`` 
     - | The index to use to scan for documents to count. 
       | Default: ``nil``

   * - ``Limit`` 
     - | The maximum number of documents to count. 
       | Default: ``0`` 

   * - ``MaxTime``
     - | The maximum amount of time that the query can run on the server.
       | Default: ``nil``

   * - ``Skip`` 
     - | The number of documents to skip before counting.
       | Default: ``0``

.. tip::

   When you use ``CountDocuments()`` to return the total number of documents in a
   collection, MongoDB performs a collection scan. You can avoid a collection scan and
   improve the performance of this method by using a hint to take advantage of the built-in index on
   the ``_id`` field. Use this technique only when calling ``CountDocuments()``
   with an empty query parameter.

   .. code-block:: csharp

>       :emphasize-lines: 1, 3

Reviewer: Q: why is the filter highlighted?

Author: To emphasize the use of an Empty filter, but I can remove if it's distracting from the point.

      var filter = Builders<Student>.Filter.Empty;

      CountOptions opts = new CountOptions(){Hint = "_id_"};
      var counter = collection.CountDocuments(filter, opts);

.. _csharp-estimated-count:

Estimated Count
---------------

> To estimate the number of documents in your collection without the use
> of a filter, use the
> ``EstimatedDocumentCount()`` method. 

Reviewer: ```suggestion
To estimate the total number of documents in your collection, use the
``EstimatedDocumentCount()`` method. 
```

.. note:: 

    The ``EstimatedDocumentCount()`` method is quicker than the
    ``CountDocuments()`` method because it uses the collection's
    metadata rather than scanning the entire collection. 

Modify Behavior
~~~~~~~~~~~~~~~

You can modify the behavior of ``EstimatedDocumentCount()`` by passing
in an ``EstimatedDocumentCountOptions`` type. If you don't specify any
options, the driver uses its default values.

The ``EstimatedDocumentCountOptions`` type allows you to configure

> options with the following property:

Reviewer: Change the language to match changes to the preceding section

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``MaxTime``
     - | The maximum amount of time that the query can run on the server.
       | Default: ``nil``

Example
```````

The following example estimates the number of documents in the
``grades`` collection:

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var count = collection.EstimatedDocumentCount();
      String output = String.Format("Estimated number of documents in
      the grades collection: " + count);

>       Console.WriteLine(output);

Reviewer: S: Combine into one line

   .. output::
      :language: none
      :visible: false

      Estimated number of documents in the grades collection: 6

.. _csharp-count-aggregation:

Aggregation
-----------

You can also include the :manual:`$count </reference/operator/aggregation/count/>`

> stage to count the number of documents in an aggregation pipeline.

Reviewer: Issue: Instead of linking to the `$count` documentation, mention that this method uses the `Count()` builder method. In the additional information section, you could link to the Server documentation

Example
~~~~~~~

The following example performs the following actions:

- Counts the number of documents where the ``finalGrade`` is greater than ``80``

> - Assigns the count to the ``count`` field

Reviewer: I: this isn't exactly what's happening in this code -- it seems like you are specifying a filter in the match stage and then appending a count stage.

.. io-code-block::
   :copyable: true

   .. input::
      :language: csharp

      var matchStage = Builders<Student>
      .Filter.Gt(f => f.finalGrade, 80);
      var pipeline = new EmptyPipelineDefinition<Student>()
      .Match(matchStage).Count();

>       Console.WriteLine(pipeline);

Reviewer: I: format so that newlined method calls are indented. Also, this code should print the result of the aggregation. LMK if you want to review this code together.
